### PR TITLE
[RFR] Allow the user to make moves

### PR DESCRIPTION
**Pull request by Luwangel**
Choose the piece to move towards the empty place.
- Using the letters (T)op, (R)ight, (B)ottom, (L)eft
- Using the arrow keys
The puzzle must update according to the piece I moved

- [x] Optimize the docker build
- [x] Create the move function
- [x] Manage errors
- [x] Write the main loop
- [x] Create the gui (display, ask user commands)

### CODE REVIEW

File: src/game.py [hunk 1]
```diff
@@ -10,3 +20,51 @@ def build_grid(size=4):
...
+def move(grid, direction):
+    empty_tile = find_empty_tile(grid)
+
+    if direction == Move.TOP:
+        new_empty_tile = [empty_tile[0] - 1, empty_tile[1]]
+    elif direction == Move.RIGHT:
+        new_empty_tile = [empty_tile[0], empty_tile[1] + 1]
+    elif direction == Move.BOTTOM:
+        new_empty_tile = [empty_tile[0] + 1, empty_tile[1]]
+    else:
```
Comment (Kmaschta) [inline src/game.py]: Nope, handle LEFT here. If direction is not one of the four, raise an exception

File: src/renderer.py [hunk 2]
```diff
@@ -1,11 +1,41 @@
...
+            msg += '(R)ight '
+        elif move == Move.BOTTOM:
+            msg += '(B)ottom '
+        else:
+            msg += '(L)eft '
+    return msg
+
+def ask_move():
+    direction = input('Choose a direction ')
+    if direction == 'T':
```
Comment (Kmaschta) [inline src/renderer.py]: You can handle miniscules too

File: src/renderer.py [hunk 3]
```diff
@@ -1,11 +1,41 @@
+from game import Move
+
+
 def welcome():
     return 'Welcome to the 15 puzzle game.'
 
 
 def show_grid(grid):
-    grid_to_show = ''
+    grid_to_show = '\n___________________\n\n'
```
Comment (jdemangeon) [inline src/renderer.py]: You can use multiply on string in python. 
Eg: '\n' + ('_' * grid) + '\n'

**Review by jdemangeon — APPROVED** [review]

**Comment by jdemangeon** [comment]
Tests fail

File: src/game.py [hunk 4]
```diff
@@ -1,3 +1,13 @@
+from enum import Enum
```
Comment (Kmaschta) [inline src/game.py]: Is this Enum really useful ?

File: src/game.py [hunk 5]
```diff
@@ -10,3 +20,57 @@ def build_grid(size=4):
...
+            x += 1
+        y += 1
+    return [-1, -1]
+
+
+def possible_moves(grid):
+    empty_tile = find_empty_tile(grid)
+    moves = []
+
+    if empty_tile == [-1, -1]:
```
Comment (Kmaschta) [inline src/game.py]: This is an impossible case

**Review by Kmaschta — COMMENTED** [review]
Can you add a constant somewhere `EMPTY_TILE = Null` ?

In this way, you can check that way: `game[x][y] is EMPTY_TILE`

File: src/game.py [hunk 6]
```diff
@@ -10,3 +20,57 @@ def build_grid(size=4):
...
+        moves.append(Move.BOTTOM)
+    if empty_tile[1] != 0:
+        moves.append(Move.LEFT)
+    return moves
+
+
+def move(grid, direction):
+    empty_tile = find_empty_tile(grid)
+
+    if direction not in possible_moves(grid) or empty_tile == [-1, -1]:
```
Comment (Kmaschta) [inline src/game.py]: Don't manage impossible case

File: src/game.py [hunk 7]
```diff
@@ -1,3 +1,13 @@
+from enum import Enum
+
+
+class Move(Enum):
+    TOP = 1
```
Comment (Kmaschta) [inline src/game.py]: You don't have to use numbers here. Use strings `TOP = 'TOP'`, it's easier to debug

File: src/game.py [hunk 8]
```diff
@@ -10,3 +20,57 @@ def build_grid(size=4):
...
+    if direction == Move.TOP:
+        new_empty_tile = [empty_tile[0] - 1, empty_tile[1]]
+    elif direction == Move.RIGHT:
+        new_empty_tile = [empty_tile[0], empty_tile[1] + 1]
+    elif direction == Move.BOTTOM:
+        new_empty_tile = [empty_tile[0] + 1, empty_tile[1]]
+    elif direction == Move.LEFT:
+        new_empty_tile = [empty_tile[0], empty_tile[1] - 1]
+    else:
+        raise Exception('The direction provided to the move function is not allowed')
```
Comment (Kmaschta) [inline src/game.py]: Use a custom exception here

File: src/test_game.py [hunk 9]
```diff
@@ -1,9 +1,29 @@
...
-from game import build_grid
+from game import Move, build_grid, possible_moves, move
 
 
 class GameTest(unittest.TestCase):
-
     def test_build_grid(self):
-        grid = [[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12], [13, 14, 15, 0]]
-        self.assertEqual(build_grid(), grid)
+        self.assertEqual(build_grid(), [[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12], [13, 14, 15, 0]])
```
Comment (Kmaschta) [inline src/test_game.py]: Can you format your matrix?

File: src/game.py [hunk 10]
```diff
@@ -10,3 +20,57 @@ def build_grid(size=4):
...
+
+    value = grid[empty_tile[0]][empty_tile[1]]
```
Comment (Kmaschta) [inline src/game.py]: Don't mutate the grid !
Create a new one

File: src/game.py [hunk 11]
```diff
@@ -1,11 +1,12 @@
-from enum import Enum
+import numpy as np
+from game_exception import MoveException, NoEmptyTileException
```
Comment (Kmaschta) [inline src/game.py]: I would prefer `from .exception`

File: src/game.py [hunk 12]
```diff
@@ -31,46 +32,45 @@ def find_empty_tile(grid):
...
-        raise Exception('The direction provided to the move function is not allowed')
-
-    value = grid[empty_tile[0]][empty_tile[1]]
-    grid[empty_tile[0]][empty_tile[1]] = grid[new_empty_tile[0]][new_empty_tile[1]]
-    grid[new_empty_tile[0]][new_empty_tile[1]] = value
+        new_x = x - 1
+        new_y = y
 
-    return grid
+    new_grid = grid.copy()
```
Comment (Kmaschta) [inline src/game.py]: :+1:

**Review by Kmaschta — CHANGES_REQUESTED** [review]
Don't forget to handle KeyboardInterrupt

File: src/game/game.py [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+        y += 1
+    raise NoEmptyTileException('The grid should contain an empty tile')
+
+
+def possible_moves(grid):
+    empty_tile = find_empty_tile(grid)
+
+    moves = []
+
+    if not empty_tile[0] == 0:
```
Comment (Kmaschta) [inline src/game/game.py]: Can you have a constant `EMPTY_TILE` ?

**Review by Kmaschta — APPROVED** [review]

**Review by Kmaschta — APPROVED** [review]

**Pull request merged**